### ISSUE TITLE
grafana dashboard example

### ISSUE THREAD

**Issue by benitogf** [tus/tusd#154]
Would someone be interested in this addition to the wiki/docs? or at least sharing some nice graph query/config, have been working on it but there's not much information available or at least not related to tusd specifically.

**Comment by benitogf** [comment]
Thanks for the enthusiasm :) my goal is to create a chart to monitor request duration for now, found that tusd doesn't have the instrumentation needed for this kind of chart so I've been investigating on how to add them, few useful resources found so far:
 
[duration collector (in node)](https://github.com/roylines/node-epimetheus/blob/master/lib/metrics.js#L26)
[doc for the instrumentation of duration in go](https://godoc.org/github.com/prometheus/client_golang/prometheus/promhttp#InstrumentHandlerDuration)
[a very nice post on tracking request duration](https://povilasv.me/prometheus-tracking-request-duration/)

So next step is to try to implement it on tusd, would a PR to add this be welcomed?

**Comment by Acconut** [comment]
> Thanks for the enthusiasm :)

We are always pleased to help people contribute to the tus ecosystem :)

> my goal is to create a chart to monitor request duration for now

What is your definition for request duration? Do you want to include all requests into this metric (GET, HEAD, POST, PATCH)? GET and HEAD request should be relatively fast but PATCH requests may range from less than a second to a few hours, so I am not sure how you would use this information. Could you elaborate more in this idea?

**Comment by benitogf** [comment]
Very good question indeed the duration of each request will not provide meaningful information about the process, nevertheless the same duration instrumentation could be applied to: 

**upload duration**

between [start](https://github.com/tus/tusd/blob/a51f5994bba6c22369023de55acf15071ee3c69b/unrouted_handler.go#L292) and [complete](https://github.com/tus/tusd/blob/a51f5994bba6c22369023de55acf15071ee3c69b/unrouted_handler.go#L308) (or fail)

**download duration**

GET requests could be instrumented directly

**Comment by benitogf** [comment]
```
  histogram := prometheus.NewHistogramVec(prometheus.HistogramOpts{
    Name: "process_duration_seconds",
    Help: "Time taken to process a file",
  }, []string{"action"})
```

```
  start := time.Now()

  // Process here
  // how do we keep the start time related to the file being processed?
  // could be an in-memory hash with uid -> startTime?

  duration := time.Since(start)
  action := "upload" // upload or download
  histogram.WithLabelValues(action).Observe(duration.Seconds())
```

**Comment by Acconut** [comment]
> how do we keep the start time related to the file being processed?
> could be an in-memory hash with uid -> startTime?

An in-memory storage is not suitable for a setup of multiple tusd instances, where requests for the same upload may be handled by different servers. Therefore you should not store this information in memory. The solution would be to add the upload start time to its metadata which can be accessed by all instances.

However, I am still very interested in this, but I haven't yet fully figured out how to properly use histograms and summaries with Prometheus. So I need some more time there :)

**Comment by benitogf** [comment]
start time in the metadata is a much better idea :)

**Comment by Acconut** [comment]
I am going to work on this next week and keep you updated.

**Comment by geor-g** [comment]
Could people who make use of Grafana share their dashboards?

**Comment by Acconut** [comment]
In the past five years we actually have started using Grafana to monitor tusd. But I am not sure how we could share some details about our dashboard? Maybe @kvz knows a good way?

**Comment by geor-g** [comment]
Could you export it, and make the Json available?

**Comment by kvz** [comment]
We have a bunch of other metrics on our dashboard as well, but perhaps we can create a tusd specific dashboard (initially in addition, not as a replacement I would say), export that, evolve it on GitHub with the community and port back patches up- & down stream?

**Comment by kvz** [comment]
I think there also is some Grafana hosted library where we could have it indexed I believe

**Comment by geor-g** [comment]
> We have a bunch of other metrics on our dashboard as well, but perhaps we can create a tusd specific dashboard (initially in addition, not as a replacement I would say), export that, evolve it on GitHub with the community and port back patches up- & down stream?

Sounds good. :+1:

**Comment by geor-g** [comment]
> I think there also is some Grafana hosted library where we could have it indexed I believe

This https://grafana.com/grafana/dashboards/ probably?

**Comment by kvz** [comment]
Yes apologies, I was on mobile earlier, but that is exactly what I mean :) After sharing with the community and processing feedback, we should put it up there too i think.

**Comment by Acconut** [comment]
> This https://grafana.com/grafana/dashboards/ probably?

That's interesting! Is there any documentation on how we can publish there?

**Comment by Acconut** [comment]
Thank you, we will try to look into this in the future
